### ISSUE TITLE
LXBatch submission does not work in ganga athena

### ISSUE THREAD

**Issue by will-cern** [ganga-devs/ganga#545]
Hi,

Just opening a ticket for problem seen when trying to submit a ganga athena job to LXBatch. Here's the reproducer (on lxplus):

setupATLAS
asetup AthAnalysisBase,2.4.9
get_files generatePRW_jobOptions.py
echo $ASG_TEST_FILE_MC > datasets.txt
lsetup ganga
ganga athena --lsf --nocompile --inPfnListFile 'datasets.txt' --split=1 generatePRW_jobOptions.py

From the logs of the failed job, I found:
`
Py:Athena            INFO including file "AthenaCommon/runbatch.py"
Traceback (most recent call last):
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallA
rea/jobOptions/AthenaCommon/runbatch.py", line 11, in <module>
    theApp.run()     # runs until theApp.EvtMax events reached
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallArea/python/AthenaCommon/AppMgr.py", line 598, in run
    sc = self.initialize()
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallArea/python/AthenaCommon/AppMgr.py", line 562, in initialize
    self.setup()
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallArea/python/AthenaCommon/AppMgr.py", line 437, in setup
    handle = self.getHandle()
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallArea/python/AthenaCommon/AppMgr.py", line 369, in getHandle
    from GaudiPython import AppMgr as GaudiAppMgr
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallArea/python/PyUtils/Helpers.py", line 72, in root6_importhook
    m = oldimporthook(name, globals, locals, fromlist, level)
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/sw/lcg/releases/LCG_81c/ROOT/6.04.14/x86_64-slc6-gcc49-opt/lib/ROOT.py", line 301, in _importhook
    return _orig_ihook( name, *args, **kwds )
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/GAUDI/v26r2p1-lcg81z/InstallArea/python/GaudiPython/__init__.py", line 18, in <module>
    from Bindings import *
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallArea/python/PyUtils/Helpers.py", line 72, in root6_importhook
    m = oldimporthook(name, globals, locals, fromlist, level)
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/sw/lcg/releases/LCG_81c/ROOT/6.04.14/x86_64-slc6-gcc49-opt/lib/ROOT.py", line 301, in _importhook
    return _orig_ihook( name, *args, **kwds )
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/GAUDI/v26r2p1-lcg81z/InstallArea/python/GaudiPython/Bindings.py", line 33, in <module>
    import Pythonizations
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/AthAnalysisBase/2.4.8/InstallArea/python/PyUtils/Helpers.py", line 72, in root6_importhook
    m = oldimporthook(name, globals, locals, fromlist, level)
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/sw/lcg/releases/LCG_81c/ROOT/6.04.14/x86_64-slc6-gcc49-opt/lib/ROOT.py", line 301, in _importhook
    return _orig_ihook( name, *args, **kwds )
  File "/cvmfs/atlas.cern.ch/repo/sw/software/AthAnalysisBase/x86_64-slc6-gcc49-opt/2.4.8/GAUDI/v26r2p1-lcg81z/InstallArea/python/GaudiPython/Pythonizations.py", line 100, in <module>
    gbl.StatusCode.SUCCESS = 1
SystemError: error return without exception set
`

**Comment by will-cern** [comment]
Anything I can help with on this issue to get it resolved?

**Comment by drmarkwslater** [comment]
@will-cern I'll push it up my priority list!

**Comment by drmarkwslater** [comment]
@will-cern Turns out after significant debugging, it was just from an old config option that should have been updated. See #668 for details!

**Comment by will-cern** [comment]
Thanks! So this is only making it in to 6.1.23 is it? When will that release be cut so we can get it on cvmfs?
Cheers
Will

**Comment by drmarkwslater** [comment]
@will-cern Shouldn't be too long as there are some pressing LHCB things to get out soon. I'll push for next week some time!

**Comment by will-cern** [comment]
Hi Mark,

Just tried testing with 6.1.24, and unfortunately I now get the worrying message:

WARNING  CMTCONFIG = x86_64-slc6-gcc49-opt, Your CMT setup is using 64 bit - are you sure you want to use 64 bit ?
ERROR    could not get Athena version. perhaps your requirements file doesn't have ATLAS_TEST_AREA
ApplicationConfigurationError: CMT could not parse correct environment ! 
 Did you start/setup ganga in the run/ or cmt/ subdirectory of your athena analysis package ? 
(consider --debug option for more information)
INFO     Stopping Job processing before shutting down Repositories
INFO     Shutting Down Ganga Repositories
INFO     Registry Shutdown

I tried adding "--debug" but it says that's an unrecognised option!

All I did for this was setup AthAnalysisBase,2.4.18 in a directory on lxplus, setup ganga on top of that, and then tried the submission command above.

Are you able to reproduce the error?

Thanks
Will

**Comment by drmarkwslater** [comment]
@will-cern I'll have a look - we're getting 6.1.25 out ASAP anyway so assuming I can find a fix, then it will be sorted soon!

**Comment by will-cern** [comment]
Thanks. let me know if you have any trouble reproducing this.

**Comment by will-cern** [comment]
So it definitely seems to be an issue when using AthAnalysisBase:

asetup AthAnalysisBase,2.4.18

This wasn't a problem with earlier ganga versions.
 Do you think we need to introduce a unit test somewhere to ensure that we never break the submission with that release? We are pushing ganga usage hard with this release now, so it will be bad if it ever breaks.

Cheers

Will

**Comment by drmarkwslater** [comment]
@will-cern So this is **technically** not my fault 😄  The only change of significance between the 6.1.23 and 6.1.24 was a bump to the panda-tools version. This was needed because Athena releases 21+ use CMake and the version Ganga was using didn't support that. However, doing this seems to have broken AthAnalysisBase setup. I'm going to see how to fix this now.

As regards testing, this is definitely where I want to be but I'm in the process of cleaning out almost 10 years of old Atlas code so I can actually get things integrated properly with the test system (basically it's the grid credential stuff that doesn't play well) and in the process remove DQ2. So expect a much slimmer and easier to maintain GangaAtlas as well as regular and extensive testing over the next few months!

**Comment by drmarkwslater** [comment]
@will-cern Found the problem! It seems when retrieving pandatools from SVN, it didn't get the latest version (or they've stopped using it) and so the version I had **thought** was 0.5.69 actually wasn't. I've fixed the tarball and I'll get Asoka to re-install it to make sure CVMFS catches up. I've also fixed the script so it grabs the version from CVMFS instead.

**Comment by will-cern** [comment]
Hi!
Just tested the lxbatch submission with this latest patch, and I can confirm the job does run (hurray!). But at the end, the ganga marks the job as failed because of this familiar:

Could not stat output data. Marking job as failed

reason. Its because the copy of the output file failed. I specified the output location as:

--outputlocation="~will/path/to/dir" --outputtype 'ATLAS'

I see in the logs this lead to the following command:

/usr/bin/rfcp my.output.root '~will/path/to/dir/my.output.root'

And it claims that the second is an invalid argument. I think two things have gone wrong here:

1. ganga put quote marks around the output location string. I think rfcp falls over when this happens.

2. ganga needs to ensure the output directory is created. This is what happens when I submit to condor: ganga makes the output directory for me. But experimenting with rfcp locally myself I find the directories are not created, and if they don't exist then rfcp wont work.

Do you agree that these two points should be fixed?

Sorry for the continued problems

Thanks
Will

**Comment by will-cern** [comment]
Actually I do see there is an attempt to mk the directory with 'rfmkdir' further up the log file, but this fails because there are also quote marks around that path:

rfmkdir -p '~will/path/to/dir'

So actually I think all of this will be fixed if the quote marks are eliminated!

Thanks
Will

**Comment by drmarkwslater** [comment]
@will-cern So this is a combination of things actually. Firstly, `rfmkdir` does indeed fail if you use the `-p` flag and try to create an **relative** path, e.g.
```
rfmkdir -p 'my/path'
```
Secondly, it doesn't expand the `~` so this doesn't work either
```
rfmkdir -p '~/my/path'
```
because if thinks of this as a relative path `./~/my/path`. 

In summary, I can put some fixes in for this in the scripts (and will do!) but in the short term, just use a fully expanded path when submitting 😄

**Comment by will-cern** [comment]
The absolute path does indeed work for me. Thanks. I can now finally advertise that we can submit jobs to lxbatch through ganga!
One follow up problem though: in my testing, I discovered that when I submitted a job, left it a long time, and then came back to my machine and did 'ganga', ganga would put the job in the failed state because it says 

WARNING job 12.0 has disappeared from the batch system

The job actually succeeded fine, so this is a bit misleading. Could this be improved so that the job is marked as successful, I guess by checking the return code in the stderr, or just statting the output files and if they exist then the job must have been good?

Thanks
Will
